### PR TITLE
core-values api implementation

### PR DESCRIPTION
**Pull request by PradipDKshirsagar**
Issue: #29

### CODE REVIEW

File: go-backend/db/core_value.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+import (
+	"context"
+	"time"
+
+	logger "github.com/sirupsen/logrus"
+)
+
+const (
+	listCoreValuesQuery    = `SELECT * FROM core_values WHERE org_id = $1 and parent_core_value_id IS NULL`
```
Comment (iamanujvrma) [inline go-backend/db/core_value.go:11]: using select * is an anti pattern
give it a read here: https://stackoverflow.com/questions/321299/what-is-the-reason-not-to-use-select

File: go-backend/db/core_value.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error while getting core values")
+		return
+	}
+
+	return
+}
+
+func (s *pgStore) ListSubCoreValues(ctx context.Context, organisationID, coreValueID int64) (coreValues []CoreValue, err error) {
```
Comment (iamanujvrma) [inline go-backend/db/core_value.go:48]: we dont need sub core values as of now

File: go-backend/db/core_value.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+func (s *pgStore) GetCoreValue(ctx context.Context, organisationID, coreValueID int64) (coreValue CoreValue, err error) {
+	err = s.db.GetContext(
+		ctx,
+		&coreValue,
+		getCoreValueQuery,
+		organisationID,
+		coreValueID,
+	)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error while getting core value")
```
Comment (iamanujvrma) [inline go-backend/db/core_value.go:74]: logging the corevalueid and organisation id would be helpful

use logger.WithFields(logger.Fields{}) to log multiple fields

File: go-backend/db/core_value.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	deleteSubCoreValueQuery = `DELETE FROM core_values WHERE org_id = $1 and parent_core_value_id = $2`
+	deleteCoreValueQuery    = `DELETE FROM core_values WHERE org_id = $1 and id = $2`
+	updateCoreValueQuery    = `UPDATE core_values SET (core_value_text, description, updated_at) =
+		($1, $2, $3) where id = $4 and org_id = $5`
+)
+
+type CoreValue struct {
+	ID                int64     `db:"id" json:"id"`
+	OrgID             int64     `db:"org_id" json:"org_id"`
+	CoreValueText     string    `db:"core_value_text" json:"core_value_text"`
```
Comment (iamanujvrma) [inline go-backend/db/core_value.go:25]: Struct name is already CoreValue, why repeat that in fields. This can be simply called as Text

Comment (iamanujvrma) [reply go-backend/db/core_value.go:25]: similar for any other attribute also

File: go-backend/db/core_value.go (hunk 3 above)
Comment (iamanujvrma) [reply go-backend/db/core_value.go:74]: similar to other instances

File: go-backend/db/core_value.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
+package db
```
Comment (iamanujvrma) [inline go-backend/db/core_value.go:1]: where are the test cases of DB methods ? these should be integration test case, which makes actual call to DB.

File: go-backend/db/db.go [hunk 6]
```diff
@@ -9,4 +9,12 @@ type Storer interface {
...
 	//GetUser(context.Context) (User, error)
 	//Delete(context.Context, string) error
+
+	// Core values
+	ListCoreValues(context.Context, int64) ([]CoreValue, error)
+	GetCoreValue(context.Context, int64, int64) (CoreValue, error)
+	CreateCoreValue(context.Context, int64, CoreValue) error
+	DeleteCoreValue(context.Context, int64, int64) error
+	UpdateCoreValue(context.Context, int64, int64, CoreValue) error
+	ListSubCoreValues(context.Context, int64, int64) ([]CoreValue, error)
```
Comment (iamanujvrma) [inline go-backend/db/db.go:19]: no need for sub core values

File: go-backend/db/core_value.go (hunk 4 above)
Comment (iamanujvrma) [reply go-backend/db/core_value.go:25]: we will be making this change in db schema too.
cc @sunil4sonawane

File: go-backend/service/core_value_http.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+func listSubCoreValuesHandler(deps Dependencies) http.HandlerFunc {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		vars := mux.Vars(req)
+		organisationID, err := strconv.ParseInt(vars["organisation_id"], 10, 64)
+		if err != nil {
+			logger.WithField("err", err.Error()).Error("Error while parsing organisation_id from url")
+			rw.WriteHeader(http.StatusBadRequest)
+			return
+		}
```
Comment (iamanujvrma) [inline go-backend/service/core_value_http.go:66]: new line

Comment (iamanujvrma) [reply go-backend/service/core_value_http.go:66]: btw we dont need handler this as of now

File: go-backend/service/core_value_http.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+func getCoreValueHandler(deps Dependencies) http.HandlerFunc {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		vars := mux.Vars(req)
+		organisationID, err := strconv.ParseInt(vars["organisation_id"], 10, 64)
+		if err != nil {
+			logger.WithField("err", err.Error()).Error("Error while parsing organisation_id from url")
+			rw.WriteHeader(http.StatusBadRequest)
+			return
+		}
```
Comment (iamanujvrma) [inline go-backend/service/core_value_http.go:106]: new line

File: go-backend/service/core_value_http.go [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+func createCoreValueHandler(deps Dependencies) http.HandlerFunc {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		vars := mux.Vars(req)
+		organisationID, err := strconv.ParseInt(vars["organisation_id"], 10, 64)
+		if err != nil {
+			logger.WithField("err", err.Error()).Error("Error while parsing organisation_id from url")
+			rw.WriteHeader(http.StatusBadRequest)
+			return
+		}
```
Comment (iamanujvrma) [inline go-backend/service/core_value_http.go:141]: new line

File: go-backend/service/response.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
+package service
+
+import (
+	"encoding/json"
+	"net/http"
+
+	logger "github.com/sirupsen/logrus"
+)
+
+type messageObject struct {
```
Comment (iamanujvrma) [inline go-backend/service/response.go:10]: can you confirm if some of these structs are not already added in some existing PRs which are merged now.

File: go-backend/service/core_value_http.go [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+					messageObject: messageObject{"Invalid core value data"},
+				},
+			})
+			return
+		}
+
+		resp, err := deps.Store.UpdateCoreValue(req.Context(), organisationID, coreValueID, coreValue)
+		if err != nil {
+			logger.WithField("err", err.Error()).Error("Error while updating core value")
+			repsonse(rw, http.StatusInternalServerError, map[string]messageObject{
```
Comment (iamanujvrma) [inline go-backend/service/core_value_http.go:196]: why are we using a map[string]messageObject to send the response here,
why cannot we directly use a struct here ?

is this as per the API doc ?

File: go-backend/service/response.go (hunk 10 above)
Comment (PradipDKshirsagar) [reply go-backend/service/response.go:10]: As these structs are used for API's response, we think it should be present in the service package only.

There is one existing struct but it is defined in db package(organization.go)

File: go-backend/service/core_value_http.go (hunk 11 above)
Comment (PradipDKshirsagar) [reply go-backend/service/core_value_http.go:196]: Yes

**Pull request merged**